src/puppetlabs/trapperkeeper/services/authorization/authorization_core.clj

```diff
@@ -0,0 +1,118 @@
+(ns puppetlabs.trapperkeeper.services.authorization.authorization-core
+  (:require [clojure.string :as str]
+            [puppetlabs.kitchensink.core :as ks]
+            [schema.core :as schema]
+            [puppetlabs.trapperkeeper.authorization.rules :as rules]
+            [puppetlabs.trapperkeeper.authorization.acl :as acl])
+  (:import (java.util.regex PatternSyntaxException)))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Constants
+
+(def required-keys
+  "Keys required in an auth rule map."
+  [:path :type])
+
+(def required-or-key
+  "At least one of these keys is required in an auth rule map."
+  [:deny :allow])
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Private
+
+(defn pprint-rule
+  [rule]
+  (str/trim (ks/pprint-to-string rule)))
+
+(schema/defn transform-config-rule-to-acl :- acl/ACL
+  [{:keys [allow deny]}]
+  (let [allow (if (string? allow) [allow] allow)
+        deny (if (string? deny) [deny] deny)]
+    (let [allow-acl (reduce #(acl/allow %1 %2) acl/empty-acl allow)
+          full-acl (reduce #(acl/deny %1 %2) allow-acl deny)]
+      full-acl)))
+
+(schema/defn transform-config-rule :- rules/Rule
+  [config-rule]
+  (let [type (name (.toLowerCase (:type config-rule)))]
+    {:type (if (= type "regex") :regex :string)
+     :path (re-pattern (:path config-rule))
+     :method :any
+     :acl (transform-config-rule-to-acl config-rule)}))
+
+(defn validate-auth-config-rule!
+  "Tests to see if the given map contains the proper data to define an auth
+  rule. Returns the provided rule if successful, otherwise throws an exception
+  with a useful error message."
+  [rule]
+  (when-not (map? rule)
+    (throw (IllegalArgumentException. "An authorization rule should be specified as a map")))
+  (let [rule-keys (keys rule)]
+    (doseq [k required-keys]
+      (when-not (some #(= k %) rule-keys)
+        (throw (IllegalArgumentException. (str "The authorization rule specified as "
+                                               (pprint-rule rule)
+                                               " does not contain a '" (name k) "' key.")))))
+    (when-not (some #(.contains required-or-key %) rule-keys)
+      (throw (IllegalArgumentException.
+               (str "Authorization rule specified as  "
+                    (pprint-rule rule)
+                    " must contain either a 'deny' or 'allow' rule.")))))
+  (when-not (string? (:type rule))
+    (throw (IllegalArgumentException.
+             (str "The type set in the authorization rule specified "
+                  "as " (pprint-rule rule) " should be a "
+                  "string that is either 'path' or 'regex'."))))
+  (let [type (.toLowerCase (:type rule))]
```

Comment:
It looks like the original intent in the tk-auth library was to have `path` be assumed as a default when `type` wasn't specified, essentially making `type` optional - https://github.com/puppetlabs/trapperkeeper-authorization/blob/1ac74458d6ee857f0fdcbf198584f635ed0ed1bd/src/puppetlabs/trapperkeeper/authorization/file/config.clj#L30.  This code would make `path` required in the rule.  In Puppet's auth.conf, the "path" expression itself controlled whether a literal or regex was used, so there wasn't really a notion of a default like there is where `type` is a separate element.  I'm okay with it being required as the code has here - makes things more explicit - and it isn't obvious to me whether literal or regex expressions would be used more often.  Figured I'd point that out in case others had thoughts on that.